### PR TITLE
Backport 2.16: Clarify that the full config enables everything that can be tested together

### PR DESCRIPTION
**Pull request by gilles-peskine-arm**
This is a partial backport of #3190.

* Copy the general guidance around `full` and `baremetal` from `config.py` to `config.pl`, but not the per-symbol comments, and of course not the code structure.
* Fix the build failures with options that are now enabled in `full`.
* Make the code strict C99.
* Reorganize the testing around deprecated features.

In the LTS branches, there are no deprecated features in `full`, so I didn't backport `full_no_deprecated` nor the corresponding `all.sh` components.

There may be some debate about exactly which parts should be backported. For some things I think that it's a close call between keeping things aligned in all branches and keeping things simple where only development needs the extra complexity (`full_no_deprecated`). So I'll wait until at least one approval on this PR to make the 2.7 backport, which I expect to be very similar to 2.16.

### CODE REVIEW

**Comment by gilles-peskine-arm** [comment]
Following our chat earlier today, I excluded `MBEDTLS_DEPRECATED_WARNING` from `full`.

File: tests/scripts/all.sh [hunk 1]
```diff
@@ -818,24 +818,27 @@ component_test_full_cmake_clang () {
...
+component_test_default_no_deprecated () {
+    msg "build: make, default + MBEDTLS_DEPRECATED_REMOVED" # ~ 30s
+    scripts/config.pl set MBEDTLS_DEPRECATED_REMOVED
+    make CC=gcc CFLAGS='-O -Werror -Wall -Wextra'
+
+    msg "test: make, default + MBEDTLS_DEPRECATED_REMOVED" # ~ 5s
+    make test
+}
+
+component_test_deprecated () {
```
Comment (gilles-peskine-arm) [inline tests/scripts/all.sh:841]: This shouldn't be here. There isn't even `MBEDTLS_TEST_DEPRECATED` in 2.16.

**Comment by gilles-peskine-arm** [comment]
I realized that I'd made some mistakes with the *`deprecated*` components of `all.sh`, so I rewrote the last two commits.

File: tests/scripts/all.sh [hunk 2]
```diff
@@ -818,25 +818,33 @@ component_test_full_cmake_clang () {
...
+    msg "build: make, default + MBEDTLS_DEPRECATED_REMOVED" # ~ 30s
+    scripts/config.pl set MBEDTLS_DEPRECATED_REMOVED
+    make CC=gcc CFLAGS='-O -Werror -Wall -Wextra'
+
+    msg "test: make, default + MBEDTLS_DEPRECATED_REMOVED" # ~ 5s
+    make test
+}
+
+component_test_full_deprecated_warning () {
+    # Test that there is nothing deprecated in the full configuraration.
```
Comment (ronald-cron-arm) [inline tests/scripts/all.sh:833]: typo: configuraration instead of configuration

Comment (gilles-peskine-arm) [reply tests/scripts/all.sh]: This was added in the last commit so I amended it.

**Pull request merged**